### PR TITLE
ini_file: add option to deal with multiple sections

### PR DESCRIPTION
**Pull request by jakob-bebop**
##### SUMMARY
ini_file: Add an optional parameter `section_has`. If the target ini file contains more than one `section`, use `section_has` to specify which one should be updated.

Use case: In wireguard config there can be multiple [Peer] sections. See example in source code doc string

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ini_file

### CODE REVIEW

File: changelogs/fragments/ini_file-section_has.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+minor_changes:
+  - "ini_file: Add an optional parameter `section_has`. If the
+    target ini file contains more than one `section`, use `section_has`
+    to specify which one should be updated."
```
Comment (felixfontein) [inline changelogs/fragments/ini_file-section_has.yml:4]:
```suggestion
  - "ini_file - add an optional parameter ``section_has``. If the
    target ini file contains more than one ``section``, use ``section_has``
    to specify which one should be updated
    (https://github.com/ansible-collections/community.general/pull/7496)."
```

File: plugins/modules/ini_file.py [hunk 2]
```diff
@@ -9,7 +9,6 @@
 # SPDX-License-Identifier: GPL-3.0-or-later
 
 from __future__ import absolute_import, division, print_function
-__metaclass__ = type
```
Comment (felixfontein) [inline plugins/modules/ini_file.py:12]: Please keep this line.

File: plugins/modules/ini_file.py [hunk 3]
```diff
@@ -46,6 +45,21 @@
...
       - Using V(null) is also required if the config format does not support sections.
     type: str
+  section_has:
+    type: list
+    elements: dict
+    required: false
+    suboptions:
+      option:
+        type: str
+        description: the option to look for within the section
```
Comment (felixfontein) [inline plugins/modules/ini_file.py:55]: Paragraphs should always consist of proper sentences.
```suggestion
        description: The option to look for within the section.
```

File: plugins/modules/ini_file.py [hunk 4]
```diff
@@ -46,6 +45,21 @@
...
+      value:
+        type: str
+        description: locate the section with this specific value
```
Comment (felixfontein) [inline plugins/modules/ini_file.py:58]:
```suggestion
        description: Locate the section with this specific value.
```

File: plugins/modules/ini_file.py [hunk 5]
```diff
@@ -46,6 +45,21 @@
...
+    description:
+      Among possibly multiple sections of the same name, select the one that contains these values.
+      - with O(state=present), if a suitable section is not found, a new section will be added, including the required options.
+      - with O(state=absent), at most one O(section) is removed if it contains the values
```
Comment (felixfontein) [inline plugins/modules/ini_file.py:62]: Linebreaks will be removed in a paragraph, so better make this three paragraphs.
```suggestion
      - Among possibly multiple sections of the same name, select the one that contains these values.
      - With O(state=present), if a suitable section is not found, a new section will be added, including the required options.
      - With O(state=absent), at most one O(section) is removed if it contains the values.
```

File: plugins/modules/ini_file.py [hunk 6]
```diff
@@ -116,14 +130,6 @@
...
     type: bool
     default: false
-  modify_inactive_option:
-    description:
-      - By default the module replaces a commented line that matches the given option.
-      - Set this option to V(false) to avoid this. This is useful when you want to keep commented example
-        C(key=value) pairs for documentation purposes.
-    type: bool
-    default: true
-    version_added: 8.0.0
```
Comment (felixfontein) [inline plugins/modules/ini_file.py:126]: This option must stay. Removing it would be a breaking change.

File: plugins/modules/ini_file.py [hunk 7]
```diff
@@ -179,16 +185,39 @@
...
     mode: '0600'
     state: present
+
+- name: remove the peer configuration for 10.128.0.11/32
+  community.general.ini_file:
+    path: /etc/wireguard/wg0.conf
+    section: Peer
+    section_has:
+      - option: AllowedIps
+      - value: 10.4.0.11/32
```
Comment (felixfontein) [inline plugins/modules/ini_file.py:195]: Shouldn't this be:
```suggestion
      - option: AllowedIps
        value: 10.4.0.11/32
```
? Same below.

File: plugins/modules/ini_file.py [hunk 8]
```diff
@@ -198,7 +227,7 @@ def match_opt(option, line):
 
 def match_active_opt(option, line):
     option = re.escape(option)
-    return re.match('()( |\t)*(%s)( |\t)*(=|$)( |\t)*(.*)' % option, line)
+    return re.match('( |\t)*(%s)( |\t)*(=|$)( |\t)*(.*)' % option, line)
```
Comment (felixfontein) [inline plugins/modules/ini_file.py:230]: It looks like you are removing a previous PR. Please undo that.

File: plugins/modules/ini_file.py [hunk 9]
```diff
@@ -471,16 +528,20 @@ def main():
         argument_spec=dict(
             path=dict(type='path', required=True, aliases=['dest']),
             section=dict(type='str'),
+            section_has=dict(type='list', elements='dict', options=dict(
+                option=dict(type='str'),
+                value=dict(type='str')
```
Comment (felixfontein) [inline plugins/modules/ini_file.py:533]: Do I understand it correctly that you can also specify only one of `option` and `value`, or even none? I would expect that I have to specify both of them (i.e. both are required).

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added some first comments.

It seems like you are reverting a previous PR as part of this PR. That's not allowed. Also please undo all reformattings of existing code.

File: plugins/modules/ini_file.py [hunk 10]
```diff
@@ -46,6 +45,21 @@
       - If left empty, being omitted, or being set to V(null), the O(option) will be placed before the first O(section).
       - Using V(null) is also required if the config format does not support sections.
     type: str
+  section_has:
+    type: list
+    elements: dict
+    required: false
```
Comment (jpmens) [inline plugins/modules/ini_file.py:51]: I think a `version_added` is missing

File: plugins/modules/ini_file.py (hunk 9 above)
Comment (jakob-bebop) [reply plugins/modules/ini_file.py:533]: I suppose they should both be required.. Can you do that by e.g. `option=dict(type='str', required=True)`?

File: plugins/modules/ini_file.py (hunk 8 above)
Comment (jakob-bebop) [reply plugins/modules/ini_file.py:230]: Oh I think I applied an old version of the file somehow. I'll start over and make a new pr.

**Comment by jakob-bebop** [comment]
moved to https://github.com/ansible-collections/community.general/pull/7505

**Pull request closed without merging**